File: csharp/OfficeCleaner/OfficeCleaner1/RobotCleaner.cs
namespace OfficeCleaner1;

class RobotCleaner
{
   private RobotCleanerParser _cleanerParser;
        private Coordinate position;
        
        // I don't have access to a HashSet, this is the next closest thing
        // using a string version means not having to write a hashing function etc. for now.
        // the int is regrettably necessary. Later versions of .NET have HashSet.
        
        private Dictionary<string, int> coordHashSet = new Dictionary<string, int>();

        // This entire block should be set up with some kind of contingency injection
        // but for now has been coded like this

        private static Direction north = new North();
        private static Direction south = new South();
        private static Direction east = new East();
        private static Direction west = new West();

        private static string northString = "N";
        private static string southString = "S";
        private static string eastString = "E";
        private static string westString = "W";

        private static List<string> moveList = new List<string>{
                northString, 
                southString,
                eastString,
                westString
            };

        private static Dictionary<string, Direction> directionTable = new Dictionary<string, Direction>
            {
                {northString, north},
                {southString, south},
                {eastString, east},
                {westString, west}
            };

        // block ends

        public RobotCleaner()
        {
            _cleanerParser = new RobotCleanerParser(moveList);
        }

        public int getVisitedPositions()
        {
            return coordHashSet.Count;
        }

        public bool parseInput(List<string> input)
        {
            coordHashSet.Clear();
            if (false == _cleanerParser.parse(input))
                return false;

            position = _cleanerParser.getStartPosition();
            coordHashSet.Add(position.ToString(), 0);

            foreach (string command in _cleanerParser.getCommands())
            {
                if (command == null)
                {
                    continue;
                }
                string[] commandParts = command.Split(' ');
                int iterations = System.Convert.ToInt32(commandParts[1]);

                Direction move = directionTable[commandParts[0]];
                for (int count = 0; count < iterations; count++)
                {
                    move.move(position);
                    if (!coordHashSet.ContainsKey(position.ToString()))
                    {
                        coordHashSet.Add(position.ToString(), 0);
                    }
                }
            }

            return true;
        }
}

File: php/src/OfficeCleaner1/RobotCleaner.php
<?php

declare(strict_types=1);

namespace OfficeCleaner1;

class RobotCleaner
{
    private $cleanerParser;
    private $position;

    // Using an associative array to mimic a HashSet for storing coordinates
    private $coordHashSet = [];

    // This entire block should be set up with some kind of dependency injection
    // but for now has been coded like this

    private static North $north;
    private static South $south;
    private static East $east;
    private static West $west;

    private static string $northString = "N";
    private static string $southString = "S";
    private static string $eastString = "E";
    private static string $westString = "W";

    private static array $moveList = [];
    private static array $directionTable = [];

    // block ends

    public function __construct()
    {
        self::$north = new North();
        self::$south = new South();
        self::$east = new East();
        self::$west = new West();

        self::$moveList = [
            self::$northString,
            self::$southString,
            self::$eastString,
            self::$westString
        ];

        self::$directionTable = [
            self::$northString => self::$north,
            self::$southString => self::$south,
            self::$eastString => self::$east,
            self::$westString => self::$west
        ];

        $this->cleanerParser = new RobotCleanerParser(self::$moveList);
    }

    public function getVisitedPositions(): int
    {
        return count($this->coordHashSet);
    }

    public function parseInput(array $input): bool
    {
        $this->coordHashSet = [];
        if (!$this->cleanerParser->parse($input)) {
            return false;
        }

        $this->position = $this->cleanerParser->getStartPosition();
        $this->coordHashSet[$this->position->__toString()] = 0;

        foreach ($this->cleanerParser->getCommands() as $command) {
            if ($command == null) {
                continue;
            }
            $commandParts = explode(' ', $command);
            $iterations = intval($commandParts[1]);

            $move = self::$directionTable[$commandParts[0]];
            for ($count = 0; $count < $iterations; $count++) {
                $move->move($this->position);
                if (!array_key_exists($this->position->__toString(), $this->coordHashSet)) {
                    $this->coordHashSet[$this->position->__toString()] = 0;
                }
            }
        }

        return true;
    }
}


